rwkv.cpp

```diff
@@ -265,194 +465,382 @@ struct ggml_tensor * rwkv_layer_norm(ggml_context * ctx, struct ggml_tensor * x,
+    // xk = x * time_mix_k + state[5 * i + 1] * (1 - time_mix_k)
+    struct ggml_tensor * xk = ggml_add_inplace(ctx,
+        ggml_mul(ctx, x0, layer.att_time_mix_k),
+        ggml_mul(ctx, layer.att_xx, rwkv_1_minus_x(ctx, layer.att_time_mix_k))
+    );
 
-            // xk = x * time_mix_k + state[5 * i + 0] * (1 - time_mix_k)
-            struct ggml_tensor * xk = ggml_add_inplace(
-                ctx,
-                ggml_mul(ctx, x0, layer.ffn_time_mix_k),
-                ggml_mul(ctx, x_prev, rwkv_1_minus_x(ctx, layer.ffn_time_mix_k))
-            );
+    // xv = x * time_mix_v + state[5 * i + 1] * (1 - time_mix_v)
+    struct ggml_tensor * xv = ggml_add_inplace(ctx,
+        ggml_mul(ctx, x0, layer.att_time_mix_v),
+        ggml_mul(ctx, layer.att_xx, rwkv_1_minus_x(ctx, layer.att_time_mix_v))
+    );
 
-            // xr = x * time_mix_r + state[5 * i + 0] * (1 - time_mix_r)
-            struct ggml_tensor * xr = ggml_add_inplace(
-                ctx,
-                ggml_mul(ctx, x0, layer.ffn_time_mix_r),
-                ggml_mul(ctx, x_prev, rwkv_1_minus_x(ctx, layer.ffn_time_mix_r))
-            );
+    // xr = x * time_mix_r + state[5 * i + 1] * (1 - time_mix_r)
+    struct ggml_tensor * xr = ggml_add_inplace(ctx,
+        ggml_mul(ctx, x0, layer.att_time_mix_r),
+        ggml_mul(ctx, layer.att_xx, rwkv_1_minus_x(ctx, layer.att_time_mix_r))
+    );
 
-            // state[5 * i + 0] = x
-            state_parts[part_index] = x0;
+    // r = torch.sigmoid(rw @ xr)
+    struct ggml_tensor * r = rwkv_sigmoid(ctx, ggml_mul_mat(ctx, layer.att_receptance, xr));
+    // k = kw @ xk
+    struct ggml_tensor * k = ggml_mul_mat(ctx, layer.att_key, xk);
+    // v = vw @ xv
+    struct ggml_tensor * v = ggml_mul_mat(ctx, layer.att_value, xv);
+
+    // ww = time_first + k
+    struct ggml_tensor * ww = ggml_add(ctx, layer.att_time_first, k);
+    // qq = torch.maximum(pp, ww)
+    struct ggml_tensor * qq = rwkv_max(ctx, layer.att_pp, ww);
+    // e1 = torch.exp(pp - qq)
+    struct ggml_tensor * e1 = rwkv_exp(ctx, ggml_sub(ctx, layer.att_pp, qq));
+    // e2 = torch.exp(ww - qq)
+    struct ggml_tensor * e2 = rwkv_exp(ctx, ggml_sub(ctx, ww, qq));
+
+    // a = e1 * aa + e2 * v
+    struct ggml_tensor * a = ggml_add_inplace(ctx, ggml_mul(ctx, e1, layer.att_aa), ggml_mul(ctx, e2, v));
+    // b = e1 * bb + e2
+    struct ggml_tensor * b = ggml_add_inplace(ctx, ggml_mul(ctx, e1, layer.att_bb), e2);
+
+    // ww = pp + time_decay
+    ww = ggml_add(ctx, layer.att_pp, layer.att_time_decay);
+    // qq = torch.maximum(ww, k)
+    qq = rwkv_max(ctx, ww, k);
+    // e1 = torch.exp(ww - qq)
+    e1 = rwkv_exp(ctx, ggml_sub(ctx, ww, qq));
+    // e2 = torch.exp(k - qq)
+    e2 = rwkv_exp(ctx, ggml_sub(ctx, k, qq));
+
+    // state[5 * i + 1] = x0
+    // state[5 * i + 2] = e1 * aa + e2 * v
+    // state[5 * i + 3] = e1 * bb + e2
+    // state[5 * i + 4] = qq
+
```

Comment:
Why people keep adding this empty line here? 🤔 This is a second or third PR from different person I've seen. This block of comments directly relates to code below, and is similar to comments above -- which do not have an empty line after them.